src/main/java/undo/UndoController.java

```diff
@@ -0,0 +1,84 @@
+package undo;
+
+import backend.Logic;
+import backend.resource.TurboIssue;
+import javafx.application.Platform;
+import javafx.geometry.Insets;
+import javafx.scene.control.Label;
+import javafx.util.Pair;
+import ui.NotificationController;
+import ui.components.Notification;
+import undo.actions.Action;
+import util.DialogMessage;
+
+import java.util.Optional;
+
+public class UndoController {
+
+    @SuppressWarnings("SpellCheckingInspection")
+    private static final String OCTICON_INFO = "\uf059";
+
+    private final Logic logic;
+    private final NotificationController notificationController;
+    private Optional<Pair<TurboIssue, Action<TurboIssue>>> undoBuffer;
+
+    // Holds an undo buffer of size 1.
+    // Buffer stores the undo action.
+    public UndoController(Logic logic, NotificationController notificationController) {
+        this.logic = logic;
+        this.notificationController = notificationController;
+        undoBuffer = Optional.empty();
+    }
+
+    // Gets called when an action is undone
+    public void undoCallback() {
+        if (undoBuffer.isPresent()) {
+            determineLogicMethod(undoBuffer.get().getKey(), undoBuffer.get().getValue());
+            undoBuffer = Optional.empty();
+        }
+    }
+
+    // Adds an action to the undo buffer.
+    @SuppressWarnings("unchecked")
+    public void addAction(TurboIssue issue, Action action) {
+        determineLogicMethod(issue, action);
+        undoBuffer = Optional.of(new Pair<>((TurboIssue) action.act(issue), action.invert()));
+        Notification notification = new Notification(createInfoOcticon(),
+                "Undo " + action.getDescription() + " for #" + issue.getId() + ": " + issue.getTitle(),
+                "Undo", this::undoCallback);
+        notificationController.showNotification(notification);
+    }
+
+    // Check to determine which logic method to call
+    @SuppressWarnings("unchecked")
+    private void determineLogicMethod(TurboIssue issue, Action action) {
+        if (action.getType() == Action.ActionType.CHANGELABELS) {
+            logic.replaceIssueLabels(issue, action)
```

Comment:
I don't really like calling `Logic` within the `Action` as I see the `Action` more as a diff rather than to encapsulate the actual method calls?

The `getType` was implemented because I couldn't get `instanceOf` to work reliably with the `Optional`. 
So you would prefer something like:
```
private void determineLogicMethod(TurboIssue issue, ChangeLabelsAction action)
```
and then a catch-all at the end?
```
private void determineLogicMethod(TurboIssue issue, Action action)
    // for unimplemented Actions?
```
would that even work?